File: leetcode/single-number.py
from typing import List


class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        # # set을 이용한 풀이
        # num_set = set()
        # for num in nums:
        #     if num in num_set:
        #         num_set.discard(num)
        #     else:
        #         num_set.add(num)
        # return num_set.pop()

        # XOR을 이용한 풀이
        answer = 0
        for num in nums:
            answer = answer ^ num
        return answer


if __name__ == '__main__':
    print(Solution().singleNumber([2,2,1]))
    print(Solution().singleNumber([4,1,2,1,2]))
    print(Solution().singleNumber([1]))

File: leetcode/SingleNumber.java
public class SingleNumber {
    public int singleNumber(int[] nums) {
        // 모든 원소가 두 번씩 등장, 유일하게 한 번만 등장하는 원소 찾기
        int answer = 0;
        for (int num : nums) {
            answer = answer ^ num;
        }
        return answer;
    }

    public static void main(String[] args) {
        int[] nums;
        
        nums = new int[]{2, 2, 1};
        System.out.println(new SingleNumber().singleNumber(nums));

        nums = new int[]{4, 1, 2, 1, 2};
        System.out.println(new SingleNumber().singleNumber(nums));
    }
}
